llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp

```diff
@@ -559,13 +618,29 @@ bool AMDGPUCallLowering::lowerFormalArgumentsKernel(
 
         B.buildAddrSpaceCast(VRegs[i][0], PtrReg);
       }
-    } else {
-      ArgInfo OrigArg(VRegs[i], Arg, i);
-      const unsigned OrigArgIdx = i + AttributeList::FirstArgIndex;
-      setArgFlags(OrigArg, OrigArgIdx, DL, F);
-      lowerParameter(B, OrigArg, ArgOffset, Alignment);
+      i++;
+      continue;
+    }
+
+    auto &PreloadKernArgs = Info->getArgInfo().PreloadKernArgs;
+    auto PreloadKernArg =
+        Arg.hasInRegAttr() ? PreloadKernArgs.find(i) : PreloadKernArgs.end();
+    if (PreloadKernArg != PreloadKernArgs.end()) {
+      lowerPreloadedParameter(B, VRegs[i], ArgTy, ArgOffset, Alignment,
+                              PreloadKernArg->getSecond().Regs);
+      ++i;
+      continue;
     }
 
+    if (Arg.hasAttribute("amdgpu-hidden-argument"))
+      F.getContext().diagnose(DiagnosticInfoUnsupported(
```

Comment:
DiagnosticInfoUnsupported is probably the wrong error kind, but this matches the DAG